src/core/common/trickle_timer.hpp

```diff
@@ -0,0 +1,180 @@
+ *
+ * @brief
+ *   This module includes definitions for the trickle timer logic.
+ *
+ * @{
+ *
+ */
+
+/**
+ * Represents the modes of operation for the TrickleTimer
+ */
+typedef enum TrickleTimerMode
+{
+    kTrickleTimerModeNormal = 0,  ///< Runs the normal trickle logic.
+    kTrickleTimerModeMPL    = 1,  ///< Runs the trickle logic modified for MPL.
+} TrickleTimerMode;
+
+/**
+ * This class implements a trickle timer.
+ *
+ */
+class TrickleTimer
+{
+public:
+    /**
+     * This function pointer is called when the timer expires.
+     *
+     * @param[in]  aContext  A pointer to arbitrary context information.
+     *
+     * @retval TRUE   If the trickle timer should continue running.
+     * @retval FALSE  If the trickle timer should stop running.
+     */
+    typedef bool (*Handler)(void *aContext);
+
+    /**
+     * This constructor creates a trickle timer instance.
+     *
+     * @param[in]  aScheduler               A refrence to the timer scheduler.
+     * @param[in]  aRedundancyConstant      The redundancy constant for the timer, k.
+     * @param[in]  aMode                    The operating mode for the timer.
+     * @param[in]  aTransmitHandler         A pointer to a function that is called when transmission should occur.
+     * @param[in]  aIntervalExpiredHandler  An optional pointer to a function that is called when the interval expires.
+     * @param[in]  aContext                 A pointer to arbitrary context information.
+     *
+     */
+    TrickleTimer(TimerScheduler &aScheduler, uint32_t aRedundancyConstant, TrickleTimerMode aMode,
+                 Handler aTransmitHandler, Handler aIntervalExpiredHandler, void *aContext);
+
+    /**
+     * This method indicates whether or not the trickle timer instance is running.
+     *
+     * @retval TRUE   If the trickle timer is running.
+     * @retval FALSE  If the trickle timer is not running.
+     */
+    bool IsRunning(void) const;
+
+    /**
+     * This method start the trickle timer.
+     *
+     * @param[in]  aIntervalMin  The minimum interval for the timer, Imin.
+     * @param[in]  aIntervalMax  The maximum interval for the timer, Imax.
+     *
+     */
+    void Start(uint32_t aIntervalMin, uint32_t aIntervalMax);
+
+    /**
+     * This method stops the trickle timer.
+     *
+     */
+    void Stop(void);
+
+    /**
+     * This method indicates to the trickle timer a 'consistent' state.
+     *
+     */
+    void IndicateConsistent(void);
+
+    /**
+     * This method indicates to the trickle timer an 'inconsistent' state.
+     *
+     */
+    void IndicateInconsistent(void);
+
+private:
+    bool TransmitFired(void) { return mTransmitHandler(mContext); }
+    bool IntervalExpiredFired(void) { return mIntervalExpiredHandler ? mIntervalExpiredHandler(mContext) : true; }
+
+    void StartNewInterval(void);
+
+    static void HandleTimerFired(void *aContext);
+    void HandleTimerFired(void);
+
+    typedef enum TricklePhase
```

Comment:
nit: including `Trickle` in names is redundant given this is part of the Trickle class.